File: programming_and_backend_development/exercises/RubyJavaScript_Challenges/Easy/anagrams.rb
=begin
Write a program that takes a word and a list of possible anagrams and selects the correct sublist that contains the anagrams of the word.

For example, given the word "listen" and a list of candidates like "enlists", "google", "inlets", and "banana", the program should return a list containing "inlets". Please read the test suite for the exact rules of anagrams.
=end

class Anagram
  def initialize(string)
    @string = string
  end

  def match(input_array)
    input_array.select do |input|
      if input.length == @string.length &&
         input.downcase.chars.sort == @string.downcase.chars.sort &&
         input.downcase != @string.downcase
         true
      else
        false
      end
    end
    # return array with anagrams of string from input_array
  end
end

#p Anagram.new("potato").match(["a", "patoto", "Potato", "potato"])

# Start time => 1:40
# End time => 1:51

File: front_end_development/LS215/anagrams.js
// Write a Function named anagram that takes two arguments: a word and an array of words. Your function should return an array that contains all the words from the array argument that are anagrams of the word argument. For example, given the word "listen" and an array of candidate words like "enlist", "google", "inlets", and "banana", the program should return an array that contains "enlist" and "inlets".

function anagram(word, list) {
  let isAnagram = (string) => {
    wordArr = [...word].sort();
    stringArr = [...string].sort();
    let matchingChar = true;
    wordArr.forEach( (char, idx) => {
      if (char !== stringArr[idx]) matchingChar = false;
    })
    return matchingChar;
  }
  return list.filter(isAnagram);
}

console.log([
  anagram('listen', ['enlists', 'google', 'inlets', 'banana']),  // [ "inlets" ]
  anagram('listen', ['enlist', 'google', 'inlets', 'banana']),   // [ "enlist", "inlets" ]
]);